module/zfs/zio.c

```diff
@@ -3232,25 +3210,19 @@ zio_write_gang_block(zio_t *pio, metaslab_class_t *mc)
 
 		zio_t *cio = zio_write(zio, spa, txg, &gbh->zg_blkptr[g],
 		    has_data ? abd_get_offset(pio->io_abd, pio->io_size -
-		    resid) : NULL, lsize, lsize, &zp,
+		    resid) : NULL, psize, psize, &zp,
 		    zio_write_gang_member_ready, NULL,
 		    zio_write_gang_done, &gn->gn_child[g], pio->io_priority,
 		    ZIO_GANG_CHILD_FLAGS(pio), &pio->io_bookmark);
 
 		zio_gang_inherit_allocator(zio, cio);
+		/*
+		 * We do not reserve for the children writes, since we already
+		 * reserved for the parent.  Unreserve though will be called
+		 * for individual children.  We can do this since we reserve
+		 * physical size, not allocated size, which might be bigger.
```

Comment:
This comment doesn't really clarify why we can safely unreserve when we didn't reserve in the first place. The amount we reserve is copies * io_size, which wouldn't seem to be enough for the data blocks.

Also nit: `s/children/child/` in the first sentence

Reply:
Adjusted the comment.